src/ffi/expression.rs

```diff
@@ -0,0 +1,399 @@
+        const UNUSED = !(Self::EQUALS.bits()
+            | Self::NOT_EQUALS.bits()
+            | Self::REGEX.bits()
+            | Self::PREFIX.bits()
+            | Self::POSTFIX.bits()
+            | Self::GREATER.bits()
+            | Self::GREATER_OR_EQUAL.bits()
+            | Self::LESS.bits()
+            | Self::LESS_OR_EQUAL.bits()
+            | Self::IN.bits()
+            | Self::NOT_IN.bits()
+            | Self::CONTAINS.bits());
+    }
+}
+
+impl From<&BinaryOperator> for BinaryOperatorFlags {
+    fn from(op: &BinaryOperator) -> Self {
+        match op {
+            BinaryOperator::Equals => Self::EQUALS,
+            BinaryOperator::NotEquals => Self::NOT_EQUALS,
+            BinaryOperator::Regex => Self::REGEX,
+            BinaryOperator::Prefix => Self::PREFIX,
+            BinaryOperator::Postfix => Self::POSTFIX,
+            BinaryOperator::Greater => Self::GREATER,
+            BinaryOperator::GreaterOrEqual => Self::GREATER_OR_EQUAL,
+            BinaryOperator::Less => Self::LESS,
+            BinaryOperator::LessOrEqual => Self::LESS_OR_EQUAL,
+            BinaryOperator::In => Self::IN,
+            BinaryOperator::NotIn => Self::NOT_IN,
+            BinaryOperator::Contains => Self::CONTAINS,
+        }
+    }
+}
+
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_OK: i64 = 0;
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_FAILED: i64 = 1;
+pub const ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL: i64 = 2;
+
+/// Validates an ATC expression against a schema.
+///
+/// # Arguments
+///
+/// - `atc`: a C-style string representing the ATC expression.
+/// - `schema`: a valid pointer to a [`Schema`] object, as returned by [`schema_new`].
+/// - `fields_buf`: a buffer for storing the fields used in the expression.
+/// - `fields_len`: a pointer to the length of `fields_buf`.
+/// - `fields_total`: a pointer for storing the total number of fields.
+/// - `operators`: a pointer for storing the bitflags representing used operators.
+/// - `errbuf`: a buffer to store any error messages.
+/// - `errbuf_len`: a pointer to the length of the error message buffer.
+///
+/// # Returns
+///
+/// An integer indicating the validation result:
+/// - `ATC_ROUTER_EXPRESSION_VALIDATE_OK` (0): Validation succeeded.
+/// - `ATC_ROUTER_EXPRESSION_VALIDATE_FAILED` (1): Validation failed; `errbuf` and `errbuf_len` will be updated with an error message.
+/// - `ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL` (2): The provided `fields_buf` is too small.
+///
+/// If `fields_buf` is non-null and `fields_len` is sufficient, this function writes the used fields to `fields_buf`,
+/// each field terminated by `\0`. It updates `fields_len` with the required buffer length and stores the total number of fields in `fields_total`.
+///
+/// If `fields_buf` is non-null but `fields_len` is insufficient, it writes the required buffer length to `fields_len`
+/// and the total number of fields to `fields_total`, then returns `ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL`.
```

Comment:
Do we even care about `fields_total` when `ATC_ROUTER_EXPRESSION_VALIDATE_BUF_TOO_SMALL`? I'm not sure what can we do with this value if `fields_buf` is truncated.

Reply:
Currently, the implementation will not write any thing if buffer is too small. It will save all fields to be writen in a vector and always calcuate required length. This could be improved once the iterator is in use.

Reply:
Done.